docs/apis/core/customfields/index.md

```diff
@@ -0,0 +1,168 @@
+
+Developer must add custom fields callbacks to the instance edit form. If the instance is "made up" (like in mod_surveybuilder), a new form has to be created with 'id' field in it that will refer to the instanceid.
+
+The following callbacks should be used in form definition, definition_after_data, validation and after form submission:
+
+```php
+$handler->instance_form_definition($mform)
+$handler->instance_form_before_set_data()
+$handler->instance_form_definition_after_data()
+$handler->instance_form_validation()
+$handler->instance_form_save($data)
+```
+
+The `instance_form_save()` method must be called after the form was saved as the `$data` parameter must have the `id` attribute.
+
+On deletion of an instance or on deletion of the whole item call:
+
+```php
+$handler->delete_instance()
+$handler->delete_all()
+```
+
+### Retrieving instances custom fields
+
+How custom fields are used depends entirely on the situation. The following handler methods will help you to retrieve custom fields values for the given instance(s):
+
+```php
+$handler->export_instance_data()
+$handler->export_instance_data_object()
+$handler->display_custom_fields_data()
+```
+
+Additional methods for advanced usage:
+
+```php
+$handler->get_instance_data()
+$handler->get_instances_data()
+$handler->get_instance_data_for_backup()
+```
+
+Method restore_instance_data_from_backup() exists in the handler class but not implemented
+
+To retrieve the list of custom fields used in the given component/area/itemid you can use:
+
+```php
+$handler->get_categories_with_fields()
+$handler->get_fields()
+```
+
+List of fields is cached in the handler and these two functions can be called multiple times.
+
+#### Example code for course custom fields
+
+This function will return all the custom fields for a given courseid:
+
+```php
+function get_course_metadata($courseid) {
+    $handler = \core_customfield\handler::get_handler('core_course', 'course');
+    // This is equivalent to the line above.
+    //$handler = \core_course\customfield\course_handler::create();
+    $datas = $handler->get_instance_data($courseid);
+    $metadata = [];
+    foreach ($datas as $data) {
+        if (empty($data->get_value())) {
+            continue;
+        }
+        $cat = $data->get_field()->get_category()->get('name');
+        $metadata[$data->get_field()->get('shortname')] = $cat . ': ' . $data->get_value();
+    }
+    return $metadata;
+}
+```
+
+### Privacy API
+
+Custom fields API does not export or delete any data because it does not know how custom fields are used, what data is considered user data and if it is considered private or shared data.
+
+Plugins that store user information in custom fields should link subsystem in their get_metadata:
+
+```php
```

Comment:
```suggestion
```php title="Plugins that store user information in custom fields should link subsystem in their get_metadata"
```